components/notebook-controller/controllers/culling_controller.go

```diff
@@ -0,0 +1,451 @@
+const DEFAULT_CLUSTER_DOMAIN = "cluster.local"
+const DEFAULT_DEV = "false"
+
+var CULL_IDLE_TIME = 0
+var ENABLE_CULLING = false
+var IDLENESS_CHECK_PERIOD = 0
+var CLUSTER_DOMAIN = ""
+var DEV = false
+
+// When a Resource should be stopped/culled, then the controller should add this
+// annotation in the Resource's Metadata. Then, inside the reconcile loop,
+// the controller must check if this annotation is set and then apply the
+// respective culling logic for that Resource. The value of the annotation will
+// be a timestamp of when the Resource was stopped/culled.
+//
+// In case of Notebooks, the controller will reduce the replicas to 0 if
+// this annotation is set. If it's not set, then it will make the replicas 1.
+const STOP_ANNOTATION = "kubeflow-resource-stopped"
+const LAST_ACTIVITY_ANNOTATION = "notebooks.kubeflow.org/last-activity"
+const LAST_ACTIVITY_CHECK_TIMESTAMP_ANNOTATION = "notebooks.kubeflow.org/last_activity_check_timestamp"
+
+const (
+	KERNEL_EXECUTION_STATE_IDLE     = "idle"
+	KERNEL_EXECUTION_STATE_BUSY     = "busy"
+	KERNEL_EXECUTION_STATE_STARTING = "starting"
+)
+
+// Each kernel of the Notebook Server has a status.
+// KernelStatus struct:
+type KernelStatus struct {
+	ID             string `json:"id"`
+	Name           string `json:"name"`
+	LastActivity   string `json:"last_activity"`
+	ExecutionState string `json:"execution_state"`
+	Connections    int    `json:"connections"`
+}
+
+// CullingReconciler : Type of a reconciler that will be culling idle notebooks
+type CullingReconciler struct {
+	client.Client
+	Log     logr.Logger
+	Scheme  *runtime.Scheme
+	Metrics *metrics.Metrics
+}
+
+func (r *CullingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("culler", req.NamespacedName)
+
+	instance := &v1beta1.Notebook{}
+	err := r.Get(context.TODO(), req.NamespacedName, instance)
+	if err != nil && apierrs.IsNotFound(err) {
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Won't check for culling when a Notebook is being culled/stopped
+	// Remove LAST_ACTIVITY_ANNOTATION and LAST_ACTIVITY_TIMESTAMP_CHECK
+	// annotations for CR objects
+	if StopAnnotationIsSet(instance.ObjectMeta) {
+		log.Info("Notebook is already stopping")
+		removeAnnotations(&instance.ObjectMeta, r.Log)
+		err = r.Update(ctx, instance)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Ensure that the underlying Notebook Pod exists
+	foundPod := &corev1.Pod{}
+	err = r.Get(ctx, types.NamespacedName{Name: instance.Name + "-0", Namespace: instance.Namespace}, foundPod)
+	if apierrs.IsNotFound(err) {
+		log.Info("Pod not found...Won't check for culling...")
+		return ctrl.Result{}, nil
```

Comment:
let's also remove the annotations, if they exist from the CR, which was part of the previous implementation
https://github.com/kubeflow/kubeflow/blob/master/components/proposals/20220121-jupyter-notebook-idleness.md#implementation-details
https://github.com/kubeflow/kubeflow/blob/master/components/notebook-controller/controllers/notebook_controller.go#L227-L250

I'm mostly thinking the scenario where the cluster is scaled down and the notebook ends up loosing its Pod (but the node of the controller is not down yet, so the controller keep running).

in this case the controller won't delete the annotation. And once the cluster is scaled back up again then the notebook will start, but the controller will stop it since it will check the previous timestamp (and there's no activity once the pod starts)

Reply:
Sure! I modified the culling-controller to remove the `culling annotation` when the underlying Pod doesn't exist, since:
1. The culling-controller is now responsible for setting this annotation
2. The culling-controller is also checking if the underlying Pod exists